src/components/AuthForm.tsx

```diff
@@ -0,0 +1,269 @@
+      setErrors((errors) => errors.filter((error) => error.fieldName !== name));
+    } catch (error: any) {
+      setErrors((errors) => [
+        ...errors.filter((err) => err.fieldName !== name),
+        { fieldName: name as keyof AuthData, message: error.message },
+      ]);
+    }
+  };
+
+  const handleSubmit = async (e: React.FormEvent) => {
+    e.preventDefault();
+    try {
+      const schema = object().shape({
+        username: !isSignIn
+          ? string().required("Username is required")
+          : string().notRequired(),
+        email: string().email("Invalid email").required("Email is required"),
+        password: string().min(6, "Password must be at least 6 characters"),
+        confirmPassword: !isSignIn
+          ? string()
+              .oneOf([authData.password], "Passwords must match")
+              .required("Confirm password is required")
+          : string().notRequired(),
+      });
+
+      await schema.validate(authData, { abortEarly: false });
+
+      const auth = getAuth(app);
+
+      if (isSignIn) {
+        await signInWithEmailAndPassword(
+          auth,
+          authData.email,
+          authData.password
+        );
+        login();
+      } else {
+        await createUserWithEmailAndPassword(
+          auth,
+          authData.email,
+          authData.password
+        );
+        await addDoc(collection(db, "users"), {
+          username: authData.username,
+          email: authData.email,
+          password: authData.password,
+        });
+
+        setIsSignIn(true);
+      }
+
+      setAuthData({
+        email: "",
+        password: "",
+        username: "",
+        confirmPassword: "",
+      });
+    } catch (error: any) {
+      if (error.code === "auth/invalid-credential") {
+        setError("Invalid email or password!");
+      } else if (error.code === "auth/too-many-requests") {
+        setError("Too many requests, try again later!");
+      } else {
+        const validationErrors: FormError[] = error.inner.map((err: any) => ({
+          fieldName: err.path as keyof AuthData,
+          message: err.message,
+        }));
+        setErrors(validationErrors);
+      }
+    }
+  };
+
+  return (
+    <form onSubmit={handleSubmit} className={classes.form}>
+      {!isSignIn && (
+        <div>
+          <input
+            type="text"
+            id="username"
+            name="username"
+            placeholder="Username"
+            value={authData.username}
+            onChange={handleInputChange}
+            onBlur={handleBlur}
+            className={classes["form-input"]}
+          />
+          {errors.map(
+            (error) =>
+              error.fieldName === "username" && (
+                <div
+                  key={error.fieldName}
+                  className={classes["form-input__error"]}
+                >
+                  {error.message}
+                </div>
+              )
+          )}
```

Comment:
two points here. 
first, for the code to be as clear as possible, lets not map inside JSX. so move this to a function call, `renderErrorMessage`. 
second, this seems like error message for every field is the same, just the fieldName changes. so lets pass it as a prop.